R/GroupSequentialDesign.R

```diff
@@ -0,0 +1,183 @@
+#' Group-sequential design
+#'
+#' [ToDo]
+#'
+#' @slot n1 stage-one sample size
+#' @slot c1f early stopping for futility boundary
+#' @slot c1e early stopping for efficacy boundary
+#' @slot n2 stage-two sample size
+#' @slot c2_pivots vector of length order giving the values of c2 at the
+#'     pivot points of the numeric integration rule [TODO: these are not available during construction]
+#' @slot x1_norm_pivots normalized pivots for integration rule (in [-1, 1])
+#' @slot weights weights of conditional score values at x1_norm_pivots for
+#'     approximating the integral over x1.
+#'
+#' @exportClass GSDesign
+setClass("GSDesign", representation(
+        n1        = "numeric",
+        c1f       = "numeric",
+        c1e       = "numeric",
+        n2        = "numeric",
+        c2_pivots = "numeric",
+        x1_norm_pivots = "numeric",
+        weights   = "numeric"
+    ),  contains = "TwoStageDesign"
+    )
+
+
+
+#' @param n1 cf. slot
+#' @param c1f cf. slot
+#' @param c1e cf. slot
+#' @param n2 cf. slot
+#' @param c2_pivots cf. slot
+#' @param x1_norm_pivots cf. slot
+#' @param weights cf. slot
+#' @param ... further optional arguments
+#'
+#' @rdname GSDesign-class
+#' @export
+GSDesign <- function(n1, c1f, c1e, n2, c2_pivots, x1_norm_pivots, weights) {
+    if (any(diff(sapply(list(c2_pivots, x1_norm_pivots, weights), length)) != 0))
+        stop("pivots and weights must all be of the same length")
+    if (any(x1_norm_pivots < -1) | any(x1_norm_pivots > 1))
+        stop("x1_norm_pivots must be in [-1, 1], is scaled automatically")
+    if (any(weights <= 0))
+        stop("weights must be positive")
+    new("GSDesign", n1 = n1, c1f = c1f, c1e = c1e, n2 = n2,
+        c2_pivots = c2_pivots, x1_norm_pivots = x1_norm_pivots, weights = weights)
+}
+
+
+
+#' @details GQDesign simply creates a TwoStageDesign object with a Gaussian
+#'     quadrature numerical integration rule.
+#'
+#' @param order order (i.e. number of pivot points in the interior of [c1f, c1e])
+#'     of the Gaussian quadrature rule to use for integration
+#'
+#' @rdname GSDesign-class
+#' @export
+GS_GQDesign <- function(n1, c1f, c1e, n2, c2_pivots, order) {
```

Comment:
Naming conventions in R are a mess ;)

CamelCase is reserved for classes and constructors though. GSDesign is construction and therefore fine. GQDesign() is already inconsistent with that definition since it is not a constructor (we do not have a class GQDesign) - my bad! 

This should be changed. Instead, the constructor should have optional arguments, i.e., either weights and normalized pivots or just the order, right?

Reply:
Hey, I'm struggling at this point. 
When I define a method "gq_design", then the signature is "numeric" in both the TwoStageDesign as the GSDesign case. How can I make the method know which class should be returned?

Reply:
Yeah, welcome to R's awkward class systems ;) In Julia ... 
Nevermind: what you actually want to dispatch on is whether n2 is a number or a vector. In R both are numerics, so the constructor cannot dispatch on the class - you will have to manually dispatch via a simple if length(n2)==1 and length(c2) > 1 (only if the two actually differ is it safe to assume that you want a Group sequential design - otherwise it could be a two-stage design with only one pivot ;) )